### PR TITLE
Czml tests improvements

### PR DESCRIPTION
**Pull request by tomaszmrugalski**
Here's my first PR, hoping to address issue #862 

Here the list of changes:

- I think I fixed the original problem with the code assuming local timezone rather than UTC. I don't have much experience with python yet and almost zero understanding of poliastro code, so if the fix is not correct, please point me in the right direction and I'll be happy to update this fix;

- improved czml_add_to_orbit check. It now really checks the exported CZML (there was assert missing in line 306/321).

- added new tests for preamble-only CZML packet. It was useful for me for debugging, but I thought it may stay. Hey, there's not such thing as too many tests;

- reordered image and show fields in billboard in expected CZML values, so the alphabetic order is accepted;

- added comments to existing tests;

- addressed almost all problems reported by pylint;

Right now all tests, except test_czml_add_orbit, which differs very slightly from the expected values (I think by millimeters). More info about this in the first comment. There's also a minor problem reported by pylint (note in second comment).

The code has been run through pylint, tox -e reformat, tox -e check.

I've tested this code on Ubuntu 19.10 and MacOS 10.15.3, both with python 3.7.5 with all dependencies installed using pip in virtualenv.

### CODE REVIEW

**Comment by tomaszmrugalski** [comment]
Here's the error I'm observing in test_czml_add_orbit:

```
E           }, {
E               "id": 0,
E               "availability": "2013-03-18T12:00:00Z/2013-03-18T23:59:35Z",
E               "position": {
E                   "epoch": "2013-03-18T12:00:00Z",
E                   "interpolationAlgorithm": "LAGRANGE",
E                   "interpolationDegree": 5,
E                   "referenceFrame": "INERTIAL",
E                   "cartesian": [
E                       0.0,
E         -             -5874061.7735,
E         ?                         ^
E         +             -5874061.7732,
E         ?                         ^
E                       20159787.8726,
E         -             40258166.1202,
E         ?                         ^
E         +             40258166.1204,
E         ?                         ^
E                       4317.5108,
E         -             -11640044.4188,
E         ?                          ^
E         +             -11640044.4185,
E         ?                          ^
E         -             17864068.2606,
E         ?                         ^
E         +             17864068.2608,
E         ?                         ^
E         -             35673719.9898,
E         ?                       - ^
E         +             35673719.9901,
E         ?                        ^^
E                       8635.0217,
E         -             -16129541.0199,
E         ?                          ^
E         +             -16129541.0197,
E         ?                          ^
E         -             13690511.4237,
E         ?                         ^
E         +             13690511.4239,
E         ?                         ^
E         -             27339319.5726,
E         ?                         -
E         +             27339319.5732,
E         ?                        +
E                       12952.5325,
E         -             -17354822.809,
E         +             -17354822.8091,
E         ?                          +
E         -             6974596.3021,
E         ?                        ^
E         +             6974596.3025,
E         ?                        ^
E         -             13927946.9767,
E         ?                         -
E         +             13927946.9776,
E         ?                        +
E                       17270.0433,
E         -             -2602428.8844,
E         ?                        ^^
E         +             -2602428.8865,
E         ?                        ^^
E         -             -2846586.9794,
E         ?                         ^
E         +             -2846586.9792,
E         ?                         ^
E         -             -5684502.8438,
E         ?                         ^
E         +             -5684502.8433,
E         ?                         ^
E                       21587.5541,
E         -             17029383.6013,
E         ?                         ^
E         +             17029383.6011,
E         ?                         ^
E         -             5939060.5902,
E         ?                       ^^
E         +             5939060.5897,
E         ?                      + ^
E         -             11860029.9442,
E         ?                        ^^
E         +             11860029.9433,
E         ?                        ^^
E                       25905.065,
E         -             16513858.1069,
E         ?                        ^^
E         +             16513858.107,
E         ?                        ^
E         -             13042152.6757,
E         ?                         ^
E         +             13042152.6754,
E         ?                         ^
E         -             26044577.071,
E         ?                        ^
E         +             26044577.0704,
E         ?                        ^^
E                       30222.5758,
E         -             12264870.9449,
E         ?                        ^^
E         +             12264870.9451,
E         ?                        ^^
E         -             17472285.3551,
E         ?                        ^^
E         +             17472285.3549,
E         ?                        ^^
E         -             34891347.6059,
E         ?                         ^
E         +             34891347.6056,
E         ?                         ^
E                       34540.0866,
E         -             6599717.4359,
E         ?                       ^^
E         +             6599717.4362,
E         ?                       ^^
E         -             19973508.888,
E         ?                        ^
E         +             19973508.8879,
E         ?                        ^^
E         -             39886175.5838,
E         ?                         ^
E         +             39886175.5836,
E         ?                         ^
E                       38857.5975,
E         -             377235.7083,
E         ?                       ^
E         +             377235.7086,
E         ?                       ^
E                       20840447.5591,
E                       41617412.1057,
E                       43175.1083,
E         -             -5874061.7735,
E         ?                         ^
E         +             -5874061.7732,
E         ?                         ^
E                       20159787.8726,
E         -             40258166.1202,
E         ?                         ^
E         +             40258166.1204,
E         ?                         ^
E                       47492.6191,
E         -             -11640044.4188,
E         ?                          ^
E         +             -11640044.4185,
E         ?                          ^
E         -             17864068.2606,
E         ?                         ^
E         +             17864068.2608,
E         ?                         ^
E         -             35673719.9898
E         ?                       - ^
E         +             35673719.9901
E         ?                        ^^
E                   ]
E               },
```

My understanding is that the errors are very small. It's 10^-3 and the units are presumably meters, so the error is in the order of single millimeter. But this way above my level of understanding. I have no idea where that data came from, how they're currently calculated (the orbital propagation is still a bit of a myster for me). Please advise.

I'd be happy to provide full logs, but I think the automated CI scripts detect the problem easily.

**Comment by tomaszmrugalski** [comment]
Here are the problems reported by pylint:

```
$ pylint poliastro/tests/test_czml.py 
************* Module test_czml
poliastro/tests/test_czml.py:650:14: E1101: Module 'astropy.units' has no 'degree' member (no-member)
poliastro/tests/test_czml.py:650:29: E1101: Module 'astropy.units' has no 'degree' member (no-member)
poliastro/tests/test_czml.py:14:0: C0411: third party import "import pytest" should be placed before "from poliastro.bodies import Mars" (wrong-import-order)
poliastro/tests/test_czml.py:18:4: C0412: Imports from package poliastro are not grouped (ungrouped-imports)
------------------------------------------------------------------
Your code has been rated at 7.97/10 (previous run: 8.31/10, -0.34)
```

I don't understand the problem with "astropy.units has no degree". This looks like a pylint issue. If there really was no degree defined there, all tests would fail.

The C0411 and C0412 warnings were added automatically by `tox -e reformat`.  Coding guidelines were clear that this command must be run, so I did that.

File: src/poliastro/tests/test_czml.py [hunk 1]
```diff
@@ -1,21 +1,32 @@
-import pytest  # noqa: E402 isort:skip
+"""
```
Comment (astrojuanlu) [inline src/poliastro/tests/test_czml.py]: Sorry, I told you to comment the tests but I don't think it's necessary to add module and function docstrings. Tests should be simple and self-explanatory as much as possible, except for comments for weird cases or important decisions

File: src/poliastro/tests/test_czml.py [hunk 2]
```diff
@@ -1,21 +1,32 @@
...
+
 import sys
 
 from astropy import units as u  # noqa: E402
 
 from poliastro.bodies import Mars
 from poliastro.examples import iss, molniya  # noqa: E402
 
+import pytest  # noqa: E402 isort:skip
```
Comment (astrojuanlu) [inline src/poliastro/tests/test_czml.py]: Different linting tools make different decisions. We are already using Flake8 (which in turn uses PyFlakes), so let's not introduce another conflicting tool (and leave whatever isort thinks it's appropriate)

File: src/poliastro/tests/test_czml.py [hunk 3]
```diff
@@ -154,8 +169,8 @@ def test_czml_add_orbit():
         ]
     },
     "billboard": {
-        "show": true,
```
Comment (astrojuanlu) [inline src/poliastro/tests/test_czml.py]: What version did you use? Notice that we are pinning czml3 to be 0.4.0 in `pyproject.toml`, but there's a more recent one with a change

File: src/poliastro/tests/test_czml.py [hunk 4]
```diff
@@ -303,11 +318,15 @@ def test_czml_add_orbit():
     )
     extractor.add_orbit(iss, rtol=1e-4, label_text="ISS", path_show=False)
 
-    repr(extractor.packets) == expected_doc
+    assert repr(extractor.packets) == expected_doc
```
Comment (astrojuanlu) [inline src/poliastro/tests/test_czml.py]: Good catch, thanks!

**Comment by astrojuanlu** [comment]
Also, thanks a lot @tomaszmrugalski for this pull request! :raised_hands:

File: src/poliastro/tests/test_czml.py (hunk 1 above)
Comment (tomaszmrugalski) [reply src/poliastro/tests/test_czml.py]: pylint was complaining about missing docstrings. Do you want them to be removed?

File: src/poliastro/tests/test_czml.py (hunk 3 above)
Comment (tomaszmrugalski) [reply src/poliastro/tests/test_czml.py]: I didn't use conda to install anything. I was following the coding guidelines. One of the steps was to install flit and it was adding and removing packages. I don't know this tool, so don't understand what was happening. pip claims czml3 is 0.4.0.

File: src/poliastro/tests/test_czml.py (hunk 1 above)
Comment (astrojuanlu) [reply src/poliastro/tests/test_czml.py]: yes please! if anything, for consistency with the rest (I prefer removing this than having docstrings everywhere in the tests)

File: src/poliastro/tests/test_czml.py (hunk 3 above)
Comment (astrojuanlu) [reply src/poliastro/tests/test_czml.py]: I see! the order change here shouldn't be necessary then, as far as I understand, right? or were these tests failing too?

Comment (tomaszmrugalski) [reply src/poliastro/tests/test_czml.py]: The czml_add_orbit test failed with and without this change. Without this change, there was additional error complaining about the difference:

```
E               "billboard": {
E         -         "show": true,
E         -         "image": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAADsMAAA7DAcdvqGQAAADJSURBVDhPnZHRDcMgEEMZjVEYpaNklIzSEfLfD4qNnXAJSFWfhO7w2Zc0Tf9QG2rXrEzSUeZLOGm47WoH95x3Hl3jEgilvDgsOQUTqsNl68ezEwn1vae6lceSEEYvvWNT/Rxc4CXQNGadho1NXoJ+9iaqc2xi2xbt23PJCDIB6TQjOC6Bho/sDy3fBQT8PrVhibU7yBFcEPaRxOoeTwbwByCOYf9VGp1BYI1BA+EeHhmfzKbBoJEQwn1yzUZtyspIQUha85MpkNIXB7GizqDEECsAAAAASUVORK5CYII="
E         +         "image": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAADsMAAA7DAcdvqGQAAADJSURBVDhPnZHRDcMgEEMZjVEYpaNklIzSEfLfD4qNnXAJSFWfhO7w2Zc0Tf9QG2rXrEzSUeZLOGm47WoH95x3Hl3jEgilvDgsOQUTqsNl68ezEwn1vae6lceSEEYvvWNT/Rxc4CXQNGadho1NXoJ+9iaqc2xi2xbt23PJCDIB6TQjOC6Bho/sDy3fBQT8PrVhibU7yBFcEPaRxOoeTwbwByCOYf9VGp1BYI1BA+EeHhmfzKbBoJEQwn1yzUZtyspIQUha85MpkNIXB7GizqDEECsAAAAASUVORK5CYII=",
E         ?                                                                                                                                                                                                                                                                                                                                                                                                                                                                      +
E         +         "show": true
E               },
```

I can revert this and the other change that changed the show/image order if you want.

File: src/poliastro/tests/test_czml.py (hunk 1 above)
Comment (tomaszmrugalski) [reply src/poliastro/tests/test_czml.py]: Removed.

File: src/poliastro/tests/test_czml.py (hunk 2 above)
Comment (tomaszmrugalski) [reply src/poliastro/tests/test_czml.py]: Sure. Rebased and removed unnecessary commits.

Comment (astrojuanlu) [reply src/poliastro/tests/test_czml.py]: Notice that there's another `import pytest` in line 1, this is making the quality check fail

Comment (astrojuanlu) [reply src/poliastro/tests/test_czml.py]: `src/poliastro/tests/test_czml.py:9:1: F811 redefinition of unused 'pytest' from line 1`

File: src/poliastro/tests/test_czml.py (hunk 3 above)
Comment (astrojuanlu) [reply src/poliastro/tests/test_czml.py]: Oh sorry, this was the test that lacked the `assert`. Let's keep it this way for now!

**Comment by tomaszmrugalski** [comment]
Hey @astrojuanlu! Thanks for this insanely fast responses. I did not expect it :) Don't worry, I won't disappear if you respond a bit later.

Thanks for resolving the other discussions. It seems the only remaining problem is the test_czml_add_orbit. Tox seems to be [reporting the same failure (line 609)](https://dev.azure.com/poliastro/poliastro/_build/results?buildId=9&view=logs&jobId=1117f030-3e3a-5822-cfd7-97e776c1f4cb&j=1117f030-3e3a-5822-cfd7-97e776c1f4cb&t=af77bb0c-2bf7-50ca-9891-2682490c119e) as [I saw on my system](https://github.com/poliastro/poliastro/pull/872#issuecomment-600931136).

I'd be happy to update the expected string to match the exact values. Alternatively, I could come up with a comparison function that would allow some small deltas. This would be more flexible, but probably more suited for a separate PR (and some people dislike the concept altogether). Please advise what the next steps should be.

**Comment by tomaszmrugalski** [comment]
Was looking around and it seems this PR also fixes #784 (although it's unclear whether the problem still occurs or not. The original reporter encountered a seemingly unrelated problem.)

**Comment by astrojuanlu** [comment]
> I'd be happy to update the expected string to match the exact values

Yes, let's go that path then!

Also remember that the quality check is still failing because of the double `pytest` import :)

**Comment by astrojuanlu** [comment]
> Thanks for this insanely fast responses. I did not expect it :) Don't worry, I won't disappear if you respond a bit later.

Hah, don't worry! I've been a bit absent for a few weeks, and I wanted to get more involved again now that it's GSOC season :)

**Comment by astrojuanlu** [comment]
> Was looking around and it seems this PR also fixes #784 (although it's unclear whether the problem still occurs or not)

As far as I understand, we already fixed that issue in https://github.com/poliastro/poliastro/pull/783. But perhaps we fixed it in the wrong way?

**Comment by tomaszmrugalski** [comment]
Doh! The unit-tests passed everywhere, except [on windows](https://dev.azure.com/poliastro/poliastro/_build/results?buildId=12&view=logs&j=f32fc568-3d58-5c2f-5c7e-c01e8366ccfc&t=4340fcf2-110e-5f6e-b978-eb43f9f7ccae&l=1783) :(

Any chance to rerun the tests with -vv? The current output is too short (see note in line 1457) to figure out what exactly is wrong.

**Comment by astrojuanlu** [comment]
Oh, I see - this is the `tox` invocation that Azure Pipelines does:

```
tox -e py38,py38_slow -- --junitxml=junit\test-results.xml --cov-report=xml:d:\a\1\s\coverage.xml
```

based on the template we're using I suppose:

https://github.com/OpenAstronomy/azure-pipelines-templates/blob/c5d7b420fa37223e02fcc3a0aacca86a86d99157/run-tox-env.yml#L41-L44

Let's momentarily change that, we can revert later.

**Comment by astrojuanlu** [comment]
@tomaszmrugalski I committed directly to your branch

**Comment by astrojuanlu** [comment]
Oh wow... the test results were generated in Windows, and now that's thy the test is failing there.

**Comment by astrojuanlu** [comment]
Not sure what's the best way to proceed here:

1. Spend time figuring out why Windows gives a different result. Could be a software thing (how NumPy is compiled perhaps?) or it could be a hardware/platform thing...
2. Go and write function to do approximate comparisons
3. Mark test as xfail on Windows

I don't like (1) and (2) to be honest - perhaps we could go for (3) and call it a day?

**Comment by astrojuanlu** [comment]
Green, finally :green_heart: Let's get this merged! Thanks a lot @tomaszmrugalski :muscle:

**Pull request merged**